src/lib/validationRules/index.ts

```diff
@@ -0,0 +1,162 @@
+// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+//
+
+import defaultMaxAgeRule from "./validDefaultMaxAge/validDefaultMaxAge";
+import refreshTokenRule from "./refreshTokenRule/refreshTokenRule";
+import validRedirectUris from "./validRedirectUris/validRedirectUris";
+import validContext from "./validContext/validContext";
+import decentClientName from "./decentClientName/decentClientName";
+import validUriFields from "./validUriFields/validUriFields";
+import sameDomainForRedirectUris from "./sameDomainForRedirectUris/sameDomainForRedirectUris";
+import noLocalAndRemoteRedirectUris from "./noLocalAndRemoteRedirectUris/noLocalAndRemoteRedirectUris";
+import validResponseType from "./validResponseType/validResponseType";
+import validApplicationType from "./validApplicationType/validApplicationType";
+import rightAuthenticationMethod from "./rightAuthenticationMethod/rightAuthenticationMethod";
+import noUnknownFields from "./noUnknownFields/noUnknownFields";
+import validScope from "./validScope/validScope";
+import validGrantTypes from "./validGrantTypes/validGrantTypes";
+import noUnsetClientUri from "./noUnsetClientUri/noUnsetClientUri";
+
+import { ValidationResults, ValidationRule } from "../types";
+
+export const offlineRules: ValidationRule[] = [
+  decentClientName,
+  validUriFields,
+  validContext,
+  validRedirectUris,
+  refreshTokenRule,
+  validGrantTypes,
+  validScope,
+  noUnknownFields,
+  defaultMaxAgeRule,
+  validApplicationType,
+  rightAuthenticationMethod,
+  validResponseType,
+  sameDomainForRedirectUris,
+  noLocalAndRemoteRedirectUris,
+  noUnsetClientUri,
+  // We omit localization validation as it is not supported by ESS or NSS (as of 2022-08-17) but leave it here for spec-compliance guidance.
+];
+
+export async function validateDocument(
+  jsonDocument: string,
+  rules: ValidationRule[]
+): Promise<ValidationResults> {
+  // try parsing first..
+  let clientIdDocument = {};
+  try {
+    clientIdDocument = JSON.parse(jsonDocument);
+  } catch {
+    return [
+      {
+        rule: {
+          type: "local",
+          name: "Document must be valid JSON",
+          description: "The document must be a valid JSON string.",
+        },
+        status: "error",
+        title: "Invalid JSON",
+        description: "The document could not be parsed to JSON.",
+        affectedFields: [],
+      },
+    ];
+  }
+
+  const validationPromises = rules.map(async (rule) => {
+    const results = await rule.check({ document: clientIdDocument });
+    return results.map((result) => ({
+      rule: rule.rule,
+      ...result,
+    }));
+  });
+
+  return (await Promise.all(validationPromises)).flat();
+}
+
+export async function requestRemoteValidation(
+  documentIri: string
+): Promise<ValidationResults> {
+  let response: Response;
+  try {
+    response = await await fetch(
+      `/api/validate-remote-document?documentIri=${documentIri}`,
+      {
+        method: "POST",
+      }
+    );
```

Comment:
```
response = await fetch(...)
data = response.json()
```

Both can be in this try/catch as if our own API does not return JSON, then it means something went _very_ wrong indeed. Can just be the same error "Our servers could not fetch or validate your remote document at this time"